hoomd/mpcd/ReverseNonequilibriumShearFlow.cc

```diff
@@ -0,0 +1,371 @@
+    {
+    if (new_num_swap > m_num_swap)
+        {
+        GPUArray<Scalar2> swap_queue(2 * new_num_swap, m_exec_conf);
+        m_swap_queue.swap(swap_queue);
+        }
+    m_num_swap = new_num_swap;
+    }
+
+//! Set the target momentum
+void mpcd::ReverseNonequilibriumShearFlow::setTargetMomentum(Scalar p_target)
+    {
+    m_p_target = p_target;
+    }
+
+/*!
+ * \param slab_width Slab thickness
+ */
+void mpcd::ReverseNonequilibriumShearFlow::setSlabWidth(Scalar slab_width)
+    {
+    m_slab_width = slab_width;
+    requestUpdateSlabs();
+    }
+
+void mpcd::ReverseNonequilibriumShearFlow::setSlabs()
+    {
+    const BoxDim& global_box = m_pdata->getGlobalBox();
+    if (m_slab_width > Scalar(0.5) * global_box.getL().y)
+        {
+        throw std::runtime_error("Slab thickness cannot be larger than Ly/2");
+        }
+
+    const Scalar3 global_lo = global_box.getLo();
+    m_lo_pos = make_scalar2(global_lo.y, global_lo.y + m_slab_width);
+    m_hi_pos = make_scalar2(Scalar(0.0), m_slab_width);
+    }
+
+/*!
+ * \param timestep Current time step of the simulation
+ */
+void mpcd::ReverseNonequilibriumShearFlow::update(uint64_t timestep)
+    {
+    // if slabs have changed, update them
+    if (m_update_slabs)
+        {
+        setSlabs();
+        m_update_slabs = false;
+        }
+
+    findSwapParticles();
+    stageSwapParticles();
+    swapParticleMomentum();
+    }
+
+/*!
+ * Finds all particles in the "lower" and "upper" slab in y direction and
+ * puts them into two GPUArrays, sorted by their momentum closest to -/+ p_target in x direction.
+ */
+void mpcd::ReverseNonequilibriumShearFlow::findSwapParticles()
+    {
+    // fill the layers with (unsorted) particles. this uses do loop for reallocation
+    bool filled = false;
+    do
+        {
+        const size_t num_layer_lo_alloc = m_layer_lo.getNumElements();
+        const size_t num_layer_hi_alloc = m_layer_hi.getNumElements();
+            {
+            ArrayHandle<Scalar4> h_pos(m_mpcd_pdata->getPositions(),
+                                       access_location::host,
+                                       access_mode::read);
+            ArrayHandle<Scalar4> h_vel(m_mpcd_pdata->getVelocities(),
+                                       access_location::host,
+                                       access_mode::read);
+
+            ArrayHandle<Scalar2> h_layer_lo(m_layer_lo,
+                                            access_location::host,
+                                            access_mode::overwrite);
+            ArrayHandle<Scalar2> h_layer_hi(m_layer_hi,
+                                            access_location::host,
+                                            access_mode::overwrite);
+
+            // filter particles into their slab in y-direction and record momentum in x-direction
+            m_num_layer_lo = 0;
+            m_num_layer_hi = 0;
+            for (unsigned int j = 0; j < m_mpcd_pdata->getN(); j++)
```

Comment:
In C++, the idiom is to prefix increment counters in for loops. Please also replace `j` by `idx` in this loop to make the meaning of this variable clearer.
```suggestion
            for (unsigned int idx = 0; idx < m_mpcd_pdata->getN(); ++idx)
```